csrc/scheduler/mma_utils.cpp

```diff
@@ -1826,6 +1831,236 @@ class MatmulTranslator : public OptInDispatch {
+                graph.toGroup(id));
+            if (it == input_alloc_groups.end()) {
+              return true;
+            }
+            size_t input_pos = std::distance(input_alloc_groups.begin(), it);
+            if (input_pos < max_input_pos) {
+              // When we look up the position of this group in the input
+              // allocation domain, that position should not decrease as we move
+              // through tv's allocation domain. If it does, that indicates a
+              // reordering.
+              return true;
+            }
+            max_input_pos = input_pos;
+          }
+          return false;
+        };
+
+    // For each operand
+    // 1. find the fusion input
+    // 2. broadcast to create missing dims
+    // 3. permute into ...BMNK
+    // 4. set all intermediate tensors to Global
+    // 5. set all intermediate tensor allocation domains to match input's
+    // 6. if already cached, cache the output
+    //
+    // After that, create the MmaOp
+    const auto process_operand = [&](TensorView* orig_operand) {
+      // It is possible to have a horizontal fusion of two MatmulPatterns that
+      // share one operand. In these cases, we might wind up translating the
+      // same operand twice. We want to re-use the same tensor for such cases.
+      const std::vector<Val*> inputs = InputsOf::output(orig_operand);
+      NVF_ERROR(inputs.size() == 1);
+      auto* fusion_input = inputs.front()->as<TensorView>();
+      // TODO: should we verify that all ops between fusion_input and
+      // orig_operand are trivial?
+
+      // Save the fusion input's allocation domain so we can set the
+      // intermediate TVs' allocation domains and contiguity to match.
+      std::vector<ValGroup> input_alloc_groups;
+      input_alloc_groups.reserve(
+          fusion_input->getMaybeAllocationDomain().size());
+      for (IterDomain* id : fusion_input->getMaybeAllocationDomain()) {
+        input_alloc_groups.push_back(graph.toGroup(id));
+      }
+
+      // First check whether the original operand already has all output logical
+      // groups and has an allocation domain that is compatible with the fusion
+      // input. If so then we should not need to replace it.
+      if (!needs_replacement(orig_operand, input_alloc_groups)) {
+        return orig_operand;
+      }
+
+      // Build a mapping from ValGroup to position. We will update this
+      // mapping when we broadcast then use it to perform a permute if
+      // necessary.
+      std::unordered_map<ValGroup, size_t> group_position;
+      for (size_t pos : c10::irange(orig_operand->getLogicalDomain().size())) {
+        IterDomain* operand_id = orig_operand->getLogicalDomain().at(pos);
+        group_position[graph.toGroup(operand_id)] = pos;
+      }
+
+      // Reorders allocation domain and sets memory type for new TVs we
+      // introduce
+      const auto set_up_intermediate = [&input_alloc_groups,
```

Comment:
```suggestion
      const auto match_intermediate_with_fusion_input = [&input_alloc_groups,
```

That ^^^ name maybe is too verbose.